### DISCUSSION TITLE
Dynamic Reflector support

### DISCUSSION THREAD

**Discussion by night-crawler** [kube-rs/kube#1135] [General]
Hi.

Currently, I'm trying to create a dynamic reflector for a `CRD`. The problem is that I don't know CRD's GVK in compile time, so I can't create a struct the usual way.

Here comes the next approach with creating a reflector of a DynamicObject. The problem here is that `reflector::store()` requires `K` implementing the `Default` trait, though `discovery::ApiResource` does not implement it at the moment. The snippet below does not compile, though after patching `ApiResource` derives with `Default`, it does.

```rust
pub async fn crd_store() -> Result<()> {
    let client = Client::try_default().await?;

    let group = "stable.example.com";
    let version = "v1";
    let kind = "CronTab";

    let gvk = GroupVersionKind::gvk(group, version, kind);
    let (ar, caps) = discovery::pinned_kind(&client, &gvk).await?;

    let api = Api::<DynamicObject>::all_with(client, &ar);

    let lp = ListParams::default();

    // the trait bound `ApiResource: Default` is not satisfied
    // the trait `Default` is not implemented for `ApiResource`
    let (reader, writer) = reflector::store();
    let rf = reflector(writer, watcher(api, lp));

    tokio::spawn(async move {
        loop {
            let nodes = reader.state().iter().map(|r| r.name_any()).collect::<Vec<_>>();
            println!("Current {} nodes: {:?}", nodes.len(), nodes);
            tokio::time::sleep(std::time::Duration::from_secs(10)).await;
        }
    });

    let mut rfa = rf.applied_objects().boxed();
    while let Some(event) = rfa.try_next().await? {
        if caps.scope == Scope::Cluster {
            println!("saw {}", event.name_any());
        } else {
            println!("saw {} in {}", event.name_any(), event.namespace().unwrap());
        }
    }

    Ok(())
}
```

Is there any other way to create a dynamic reflector without patching? If no, can `Default` be added to the `ApiResource` implementation?

Many thanks.

**Comment by clux** [comment]
Hey there! The `reflector::store()` fn is only a helping way to construct a `Writer` and a `Store` from this writer using `Writer::default`. The `Writer::default` call does genuinely require the default bound on the dynamic type - and as such only helps out the static cases - but you __can__ create a `Writer` directly `Writer::new(dyntype)` using your `ApiResource` as a dyntype.
